cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/ETLStage.java

```diff
@@ -77,6 +104,117 @@ public ETLStage upgradeStage(UpgradeContext upgradeContext) {
     return new io.cdap.cdap.etl.proto.v2.ETLStage(name, etlPlugin);
   }
 
+  /**
+   * Updates stage by performing update action logic provided in context.
+   * Current relevant update actions for stages are:
+   *  1. UPGRADE_ARTIFACT: Upgrades plugin artifact by finding the latest version of plugin to use.
+   *
+   * @param updateContext Context to use for updating stage.
+   * @return new (updated) ETLStage.
+   */
+  public ETLStage updateStage(ApplicationUpdateContext updateContext) throws Exception {
+    for (ApplicationConfigUpdateAction updateAction: updateContext.getUpdateActions()) {
+      switch (updateAction) {
+        case UPGRADE_ARTIFACT:
+          return new io.cdap.cdap.etl.proto.v2.ETLStage(name, upgradePlugin(updateContext), inputSchema,
+                                                        outputSchema);
+        default:
+          return this;
+        }
+      }
+
+    // No update action provided so return stage as is.
+    return this;
+  }
+
+  /**
+   * Upgrade plugin used in the stage.
+   * 1. If plugin is using fixed version and a new plugin artifact is found with higher version in SYSTEM scope,
+   *    use the new plugin.
+   * 2. If plugin is using a plugin range and a new plugin artifact is found with higher version in SYSTEM scope,
+   *    move the upper bound of the range to include the new plugin artifact. Also change plugin scope.
+   *    If new plugin is in range, do not change range. (Note: It would not change range even though new plugin is in
+   *    different scope).
+   *
+   * @param updateContext To use helper functions like getPluginArtifacts.
+   * @return Updated plugin object to be used for the udated stage. Returned null if no changes to current plugin.
+   */
+  private ETLPlugin upgradePlugin(ApplicationUpdateContext updateContext) throws Exception {
+    // Currently tries to find latest plugin in SYSTEM scope and upgrades current plugin if version is higher,
+    // ignoring current plugin scope.
+    // In future, we can modify logic to fetch the latest plugin in any scope.
+    List<ArtifactId> candidates =
+      updateContext.getPluginArtifacts(plugin.getType(), plugin.getName(),
+                                       ArtifactScope.SYSTEM, null);
+    if (candidates.isEmpty()) {
+      return plugin;
+    }
+
+    // getPluginArtifacts returns plugins sorted in ascending order.
+    // TODO: Consider passing sort order as parameter.
```

Comment:
seems like something we should do now (or change the behavior to return in descending order with a limit), since there isn't any use case for getting in ascending order